node/src/braid/mod.rs

```
@@ -118,32 +157,860 @@ impl Braid {
+                );
+                let work_summation = sub_descendants
+                    .get(&bead)
+                    .map(|descendants| descendants.iter().map(|d| &bead_work[d]).sum())
+                    .unwrap_or(BigUint::zero());
+                ret_val.insert(
+                    *bead,
+                    previous_work.clone() + &bead_work[bead] + work_summation,
+                );
+            }
+            let work_summation: BigUint = curr_cohort.iter().map(|bead| &bead_work[bead]).sum();
+            previous_work += work_summation;
+        }
+        return ret_val;
+    }
+    /// Comparator function for ordering bead indices in Braidpool consensus logic.
+    ///
+    /// The comparison follows a strict priority:
+    /// 1. **Descendant Work** (`dwork`)
+    /// 2. **Ancestor Work** (`awork`)
+    /// 3. **Bead Index** (in reverse; i.e., smaller index wins last)
+    ///
+    /// This comparator is designed to be used in sorting and priority queues where
+    /// consensus-based ordering of beads is necessary (e.g., tip selection, leader election).
+    ///
+    /// # Arguments
+    ///
+    /// * `a` - Index of bead A.
+    /// * `b` - Index of bead B.
+    /// * `dwork` - Map of bead index to its **descendant work** (must be provided).
+    /// * `awork_or_not` - Optional map of bead index to **ancestor work**; required for full comparison.
+    ///
+    /// # Returns
+    ///
+    /// An `Ordering` (`Less`, `Greater`, or `Equal`) indicating the relative ranking of bead A vs B.
+    ///
+    pub fn bead_cmp(
+        a: usize,
+        b: usize,
+        dwork: &HashMap<usize, BigUint>,
+        awork_or_not: Option<&HashMap<usize, BigUint>>,
+    ) -> Result<Ordering, BraidError> {
+        if dwork[&a] < dwork[&b] {
+            return Ok(Ordering::Less);
+        }
+        if dwork[&a] > dwork[&b] {
+            return Ok(Ordering::Greater);
+        }
+
+        let awork = awork_or_not.ok_or(MissingAncestorWork)?;
```

I don't understand, why would this ever be "missing"? It might be zero (genesis) but why missing? Missing parents? (in which case we should return an error if either bead is not in the braid).

---

Yes it should ideally be the case making it non-optional since without ancestor_work existence of the valid bead will always be false hence no suitable logic behind keeping it optional .